### PR TITLE
[WPB-3633] client receives conversation deleted conversation events after federation delete event

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-3633

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/brig/src/Brig/API/Internal.hs [hunk 1]
```diff
@@ -366,13 +366,12 @@ deleteFederationRemote dom = do
...
 
 -- | Remove one-on-one conversations for the given remote domain. This is called from Galley as
--- part of the defederation process, and should not be called duriung the initial domain removal
+-- part of the defederation process, and should not be called during the initial domain removal
 -- call to brig. This is so we can ensure that domains are correctly cleaned up if a service
 -- falls over for whatever reason.
 deleteFederationRemoteGalley :: Domain -> ExceptT Brig.API.Error.Error (AppT r) ()
 deleteFederationRemoteGalley dom = do
   lift . wrapClient . Data.deleteRemoteConnectionsDomain $ dom
-  assertNoDomainsFromConfigFiles dom
```
Comment (fisx) [inline services/brig/src/Brig/API/Internal.hs:375]: the reason for this line is that if you delete a remote that's also present in the config file, it will fail silently (signal success), as read is always the union of cassandra and yaml, and yaml can't be changed here.  there is a comment about this somewhere in this CRUD family.

why did you remove it?

File: integration/test/Test/Conversation.hs [hunk 2]
```diff
@@ -177,3 +179,136 @@ testCreateConversationNonFullyConnected = do
...
+  let setFederationConfig =
+        setField "optSettings.setFederationStrategy" "allowDynamic"
+          >=> removeField "optSettings.setFederationDomainConfigs"
+          >=> setField "optSettings.setFederationDomainConfigsUpdateFreq" (Aeson.Number 1)
+  startDynamicBackends
+    [ def {dbBrig = setFederationConfig},
+      def {dbBrig = setFederationConfig}
+    ]
+    $ \dynDomains -> do
+      domains@[domainA, domainB] <- pure dynDomains
```
Comment (fisx) [inline integration/test/Test/Conversation.hs:250]: still don't understand why we have take this detour every time.  maybe it's worth fixing this in package-wide in a separate PR?

File: integration/test/Test/Conversation.hs [hunk 3]
```diff
@@ -177,3 +179,136 @@ testCreateConversationNonFullyConnected = do
...
+      connectAllDomainsAndWaitToSync 1 domains
+      [uA, uB] <- createAndConnectUsers [domainA, domainB]
+      -- figure out on which backend the 1:1 conversation is created
+      conn <- getConnection uA uB
```
Comment (fisx) [inline integration/test/Test/Conversation.hs:254]: `conn` usually refers to a cannon web socket connection of one client, but i don't have a better name either.  maybe this is fine?  (`uconn`?)

File: integration/test/Test/Conversation.hs [hunk 4]
```diff
@@ -177,3 +179,136 @@ testCreateConversationNonFullyConnected = do
...
+      when (domainA /= conversationOwningDomain && domainB /= conversationOwningDomain) $ do
+        -- this should not happen
+        assertFailure $
+          "conversation was expected to be created on "
+            <> domainA
+            <> " or "
+            <> domainB
+            <> " but was created on '"
+            <> conversationOwningDomain
+            <> "'"
```
Comment (fisx) [inline integration/test/Test/Conversation.hs:292]: we don't expect this to ever happen, right?  what domain should that be?  so i wouldn't spend so much code on an elaborate error message here, but just case-switch over `conversationOwningDomain` with a catch-all saying `error "impossible"`.

File: services/galley/src/Galley/API/Internal.hs [hunk 5]
```diff
@@ -611,50 +602,17 @@ deleteFederationDomainRemoteUserFromLocalConversations dom = do
...
-        -- to the remote backend.
-        -- We don't need to check the conversation type here, as we can't tell the
-        -- remote federation server to delete the conversation. They will have to do a
-        -- similar processing run for removing the local domain from their federation list.
-        onConversationUpdated dom convUpdate
+  remoteConvs <- foldr insertIntoMap mempty <$> E.getLocalMembersByDomain dom
+  for_ (Map.toList remoteConvs) $ \(cnv, lUsers) -> do
+    -- All errors, either exceptions or Either e, get thrown into IO
+    mapError @NoChanges (const (InternalErrorWithDescription "No Changes: Could not remove a local member from a remote conversation.")) $ do
+      E.deleteMembersInRemoteConversation (toRemoteUnsafe dom cnv) (N.toList lUsers)
```
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:615]: this is certainly more to the point!  :)

**Review by fisx — APPROVED** [review]

File: integration/test/Test/Conversation.hs (hunk 2 above)
Comment (battermann) [reply integration/test/Test/Conversation.hs:250]: I think it is alright, but sure, if we can improve it, I am fine with that.
